### ISSUE TITLE
Adding/modifying attributes of existing facts

### ISSUE THREAD

**Issue by macroach** [jruizgit/rules#92]
Hi, I'd like to be able to add attributes on facts (and ideally modify them too). For example, if a fact with just one attribute `{'val':1}` exists, then I'd like a rule's action to be able to add a new attribute to the fact, like `{'val':1,'nextval':2}`. Is this possible? (Apologies in advance if I'm just missing something!)

Here's some example code below. None of the commented lines after `def flaggedFraud(c):` cause the fact to be modified such that the `listFraud` rule subsequently fires.

```
from durable.lang import *

with ruleset('ex'):

    @when_all( pri(1),
        c.item << (
            (m.item == 'txn') &
            (-m.fraud_status)
        )
    )
    def flaggedFraud(c):
        # none of these works
        c.item.fraud_status = 'flagged'
        #c.item['fraud_status'] = 'flagged'
        #c.item.__setattr__('fraud_status','flagged')


    @when_all( pri(1),
        c.item << (
            (+m.fraud_status)
        )
    )
    def listFraud(c):
        print("{0} {1} is {2}".format(c.item.item,c.item.nmid,c.item.fraud_status))


    @when_start
    def start(host):
        host.assert_fact("ex",{"item":"txn","nmid":"1"})
        host.assert_fact("ex",{"item":"txn","nmid":"2"})
        host.assert_fact("ex",{"item":"txn","nmid":"3","fraud_status":"under investigation"})


run_all()
```

**Comment by jruizgit** [comment]
Hi, thanks for asking the question. durable_rules doesn't implement an explicit method to modify facts. Modifying facts is done by retracting the old fact and asserting the new one. For example:

```python
with ruleset('ex'):

    @when_all( pri(1),
        c.item << (
            (m.item == 'txn') &
            (-m.fraud_status)
        )
    )
    def flaggedFraud(c):
        c.retract_fact(c.item)
        c.item.fraud_status = 'flagged'
        c.assert_fact(c.item)
        

    @when_all( pri(1),
        c.item << (
            (+m.fraud_status)
        )
    )
    def listFraud(c):
        print("{0} {1} is {2}".format(c.item.item,c.item.nmid,c.item.fraud_status))


    @when_start
    def start(host):
        host.assert_fact("ex",{"item":"txn","nmid":"1"})
        host.assert_fact("ex",{"item":"txn","nmid":"2"})
        host.assert_fact("ex",{"item":"txn","nmid":"3","fraud_status":"under investigation"})
```

**Comment by macroach** [comment]
Thanks @jruizgit, that's very helpful!

I'm curious about your thoughts on whether it might be more efficient to create a single fact per attribute, with facts linked by id's. This approach makes writing rules more verbose, but since you only need to add new bitesized facts rather than retracting & adding larger facts, it _seems_ like this might be the more efficient route (as in scenarios where there are perhaps 1000 items with 100 attributes each). The flip side is that the added verbosity also means that Durable has way more facts to sort through and match in order to retrieve specific attributes.

This may be something I just need to benchmark for myself at scale, but I'd certainly be interested in hearing whether you have any initial thoughts. 

Here's an example, modifying the code sample used above:


```
from durable.lang import *


with ruleset('ex'):

    @when_all( pri(1),
        c.item << (m.item == 'txn'),
        none(
            (c.item.nmid == m.nmid) &
            (+m.fraud_status)
        )
    )
    def flaggedFraud(c):
        c.assert_fact({
            "nmid":c.item.nmid,
            "fraud_status":"flagged"
        })


    @when_all( pri(1),
        c.item << (+m.item),
        c.fraud_status << (
            (c.item.nmid == m.nmid) &
            (+m.fraud_status)
        )
    )
    def listFraud(c):
        print("{0} {1} is {2}".format(c.item.item,c.item.nmid,c.fraud_status.fraud_status))


    @when_start
    def start(host):
        host.assert_fact("ex",{"nmid":"1","item":"txn"})
        host.assert_fact("ex",{"nmid":"2","item":"txn"})
        host.assert_fact("ex",{"nmid":"3","item":"txn"})
        host.assert_fact("ex",{"nmid":"3","fraud_status":"under investigation"})


run_all()
```

**Comment by agonen** [comment]
@macroach if  you did  scale test on your idea above, can you share the results

**Comment by macroach** [comment]
Yep @agonen, it ended up being many orders of magnitude faster to do it the way @jruizgit suggested. The reason is that each time it would be joining on attributes that shared a common object id (`nmid` in my example), it had to test every micro-fact against each other, in effect doing a cartesian join (I'm admittedly making some assumptions here, and have not strictly verified the mechanics of this on the backend). 

So, suppose I have 5 objects with 10 attributes and want to find a fact having `item == 'txn'` and `status == 'pending'`. 
* If each of the 5 objects is stored as a single fact with 10 attributes, then all the rule engine has to do is scan across those 5 objects and look at those two attributes (5x2 = 10 comparisons total). If those 5 objects are stored as 50 micro-facts linked by a `obj_id` attribute (eg, `{'obj_id':1,'item':'txn'}` etc), then to match `item` and `status` together through `obj_id`, the rule engine would be checking across **1225 combinations** (`C(50,2)`).
* Now suppose we add a third condition such that `value > 500`. In this case, you'd be looking at `C(50,3)` = **39200 combinations**. 
* Adding a fourth parameter into the condition bumps it up to almost **half a million comparisons** to make. So you can see it quickly grows out of control especially as conditions become increasingly complex. Meanwhile, for that same scenario where we have a condition that looks at 4 parameters, if we have only 5 objects with 10 attributes each, the engine only needs to make 20 comparisons. 
* (_Admittedly, this is a dumbed down worst-case scenario and I'm pretty sure durable's execution is more optimized than this, but it should at least give a sense of how things scale for the two methods._) 

Anyway, on an empirical level, using micro-facts caused the rule execution to come to a veritable crawl once I got a couple dozen micro-facts and a handful of mildly complex rules (involving maybe 4-8 attributes each on average) together. On the flip side, switching over to the retract & assert method of modifying facts is very efficient and with those same facts and rules (and even more, as my rule base has grown), rules are executing at <0.05s a piece.
